src/main/java/io/jenkins/plugins/artifact_manager_jclouds/BlobStoreProvider.java

```diff
@@ -51,49 +51,83 @@
+        private final int partNumber;
+        private final String eTag;
+
+        public Part(int partNumber, String eTag) {
+            this.partNumber = partNumber;
+            this.eTag = eTag;
+        }
+
+        public int getPartNumber() {
+            return partNumber;
+        }
+
+        public String getETag() {
+            return eTag;
+        }
+    }
+
+    public interface MultipartUploader extends AutoCloseable {
+        @NonNull
+        URL toExternalURL(int partNumber) throws IOException;
+
+        void complete(@NonNull List<Part> parts) throws IOException;
+    }
+
+    /**
+     * A constant for the blob path prefix to use.
+     */
     @NonNull
     public abstract String getPrefix();
 
-    /** A constant for the blob container name to use. */
+    /**
+     * A constant for the blob container name to use.
+     */
     @NonNull
     public abstract String getContainer();
 
-    /** A constant to define whether we should delete artifacts or leave them to be managed on the blob service side. */
+    /**
+     * A constant to define whether we should delete artifacts or leave them to be managed on the blob service side.
+     */
     public abstract boolean isDeleteArtifacts();
 
-    /** A constant to define whether we should delete stashes or leave them to be managed on the blob service side. */
+    /**
+     * A constant to define whether we should delete stashes or leave them to be managed on the blob service side.
+     */
     public abstract boolean isDeleteStashes();
 
-    /** Creates the jclouds handle for working with blob. */
+    /**
+     * Creates the jclouds handle for working with blob.
+     */
     @NonNull
     public abstract BlobStoreContext getContext() throws IOException;
 
     /**
      * Get a provider-specific URI.
-     * 
-     * @param container
-     *            container where this exists.
-     * @param key
-     *            fully qualified name relative to the container.
+     *
+     * @param container container where this exists.
+     * @param key       fully qualified name relative to the container.
      * @return the URI
      */
     @NonNull
     public abstract URI toURI(@NonNull String container, @NonNull String key);
 
     /**
      * Generate a URL valid for downloading OR uploading the blob for a limited period of time
-     * 
-     * @param blob
-     *            blob to generate the URL for
-     * @param httpMethod
-     *            HTTP method to create a URL for (downloads or uploads)
+     *
+     * @param blob       blob to generate the URL for
+     * @param httpMethod HTTP method to create a URL for (downloads or uploads)
      * @return the URL
      * @throws IOException
      */
     @NonNull
     public abstract URL toExternalURL(@NonNull Blob blob, @NonNull HttpMethod httpMethod) throws IOException;
 
+    @NonNull
+    public abstract MultipartUploader initiateMultipartUpload(@NonNull Blob blob) throws IOException;
```

Comment:
Should not introduce new `abstract` methods.

Reply:
Unfortunately, I have not thought of anything better yet.
Uploading artifacts logic clearly does not fit into the existing interface. For multipart upload we need:
- check file size (worker), because we shoud known parts count and uploading time (multipart allowed only for files >= 5Mb);
- generate urls for everty part (master);
- initiate uploading to S3 (master);
- upload every part to S3 and store etag (worker);
- complete uploading to S3 (master).
Also we would like to make reduce "master <-> worker" roundtrips as possible.

Reply:
It would be more beautiful if I could request the URL to upload from worker, but:
 - I don't know how to do it;
 - it looks like as security hole.